### PR TITLE
Add support for .ts files when app is not run with ts-node.

### LINKED ISSUE #230: Typescript compiler error when running tests using Vitest
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

3.28.0

### Node.js version

16

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

12.3

### Description

I would like to test my fastify app using [vitest](https://vitest.dev), however I have noticed there is a typescript error I can't seem to get around.

When I run my tests Fastify errors, `fastify-autoload cannot import plugin ... . To fix this error compile TypeScript to JavaScript or use 'ts-node' to run your app.`. 

I can successfully run the tests when I compile them to Javascript and import the javascript version in the test. This does however, feel a bit hacky. It can also impact the code coverage (depending on how you are generating the javascript). 

### Steps to Reproduce

Stackbliz link: https://stackblitz.com/edit/node-sdwsoh

Run the `npm run test` command. 

### Expected Behavior

Tests can be run in vitest using Typescript without having to compile first, and import the javascript version.

### PR DESCRIPTION
**Pull request by kylerush**
This branch adds support for `.ts` files when the application is not run with `ts-node.` The specific use case for this is mentioned in #230. Vitest does not run with `ts-node` (it uses `es-build`) to compile TypeScript. This creates a situation where some`.ts` files that AutoLoad tries to import fails because there is no Typescript transpiler.

This pull request resolves #230.

### To make the tests fail
Replace lines 32-42 in `test/commonjs/error.js` with `app3.register(require('./index-error/app'))`. This will cause the plugin to load `ts-node` from the Node modules and not throw an error to the user that they need to install `ts-node`.

Replace `index.js` with the current version of the file in the `master` branch. This will cause the test files in `test/commonjs/ts-node.js` to error because it cannot load `.ts` files without either being run with ts-node or the ts-node `register` that is added in this branch.

## Notes
I did not modify any documentation as the error messages in `index.js` will still make sense. Happy to revise the documentation if requested.

Per request, I added `vite` and `vitest` as devDependencies and added some vitest tests so that we don't lose vitest support. 

Also happy to make any changes needed to this. Thanks for reviewing!

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin]

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -5,14 +5,14 @@ const url = require('url')
 const { readdir } = require('fs').promises
 const pkgUp = require('pkg-up')
 
-const isTsNode = (Symbol.for('ts-node.register.instance') in process) || !!process.env.TS_NODE_DEV
+// const isTsNode = (Symbol.for('ts-node.register.instance') in process) || !!process.env.TS_NODE_DEV
```
Comment (mcollina) [inline index.js:8]: remove this

File: index.js [hunk 2]
```diff
@@ -5,14 +5,14 @@ const url = require('url')
...
 const isJestEnvironment = process.env.JEST_WORKER_ID !== undefined
 const isSWCRegister = process._preload_modules && process._preload_modules.includes('@swc/register')
 const isSWCNodeRegister = process._preload_modules && process._preload_modules.includes('@swc-node/register')
 const isSWCNode = typeof process.env._ === 'string' && process.env._.includes('.bin/swc-node')
 const isTsm = process._preload_modules && process._preload_modules.includes('tsm')
 const isTsx = process._preload_modules && process._preload_modules.toString().includes('tsx')
-const typescriptSupport = isTsNode || isJestEnvironment || isSWCRegister || isSWCNodeRegister || isSWCNode || isTsm || isTsx
+// const typescriptSupport = isTsNode || isJestEnvironment || isSWCRegister || isSWCNodeRegister || isSWCNode || isTsm || isTsx
```
Comment (mcollina) [inline index.js:15]: remove this

Comment (kylerush) [reply index.js]: Removed.

File: index.js (hunk 1 above)
Comment (kylerush) [reply index.js]: Removed.

**Comment by kylerush** [comment]
Yes, will work on adding some tests with vitest. Good idea.

**Comment by kylerush** [comment]
@mcollina all set. added some tests with vitest.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by bcomnes** [comment]
This change is breaking my non-ts fastify servers https://github.com/fastify/fastify-autoload/issues/266

**Comment by ManUtopiK** [comment]
It also not work with `vite-node`. ("@fastify/autoload": "^5.4.1").
forceESM doesn't work.
```
`@fastify/autoload cannot import plugin at '${file}'. To fix this error compile TypeScript to JavaScript or use 'ts-node' to run your app.`
```

**Comment by climba03003** [comment]
Please open your own issue with a proper reproducible example.

**Pull request merged**